### DISCUSSION TITLE
Prism + DryIoC + WinUI

### DISCUSSION THREAD

**Discussion by sigmarsson** [PrismLibrary/Prism#2287] [Uno Platform]
Hello Seniors,

Can you please draw up some pieces of advice how to empower the new WinUI application with the Prism + DryIoC configuration ?

**Comment by dansiegel** [comment]
WinUI isn't directly supported, it's more indirect via Uno at the moment. You'd want to follow the Uno guidance on creating a WinUI app, and then add Prism.Uno to it.

**Reply by sigmarsson** [reply]
@dansiegel is there still an indirect support for WinUI only ? You suggested in the past to use the Prism.Uno library. May I use a WinUI desktop specific library please? 
The Uno specific package doesn't work with Mobile heads like Android as this issue exhibits.

https://github.com/unoplatform/uno/discussions/11618

**Reply by sigmarsson** [reply]
@dansiegel Thanks. The package ```Prism.Uno``` indeed works with WinUI desktop application. It does not however withe an Android head in an Uno solution. Please advise how to proceed with Prism + Uno Android project.

**Reply by dansiegel** [reply]
> @dansiegel Thanks. The package ```Prism.Uno``` indeed works with WinUI desktop application. It does not however withe an Android head in an Uno solution. Please advise how to proceed with Prism + Uno Android project. 

I'm not sure what you are having issues with as I've tested specifically with Android. Keep in mind Prism.Uno (UWP) is no longer supported and Prism.Uno.WinUI will be the only one getting updates for Prism 9

**Comment by sigmarsson** [comment]
Cool. I can't seem to find the same base class though;

`    public partial class App : Prism.PrismApplicationBase`

does not match with the base class defined in the xaml; ?

```
<prism:PrismApplication
    x:Class="WinUI.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:WinUI"
    xmlns:prism="http://prismlibrary.com">
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
            </ResourceDictionary.MergedDictionaries>
            <!--  Other app resources here  -->
        </ResourceDictionary>
    </Application.Resources>
</prism:PrismApplication>
```

**Comment by sigmarsson** [comment]
Perhaps the **PrismApplicationBase** is missing from the namespace or shall I add a different xmlns?

**Reply by brianlagunas** [reply]
You don't use PrismApplicationBase. You must use PrismApplication to match your app.xaml root element

**Comment by sigmarsson** [comment]
I thought this class was the particular one as it is located in Prism.Uno advised in the first comment. 

https://github.com/PrismLibrary/Prism/blob/master/src/Uno/Prism.Uno/PrismApplicationBase.cs

**Comment by sigmarsson** [comment]
As the Uno example advising **PrismApplication** isnt even harnessed as base class in the Ch9 demo app.

https://github.com/unoplatform/Uno.Ch9/blob/master/src/Ch9/Ch9.UWP/App.xaml.cs

**Reply by brianlagunas** [reply]
That isn't a PrismApplication and doesn't even reference Prism

**Comment by sigmarsson** [comment]
Yes, it was my mistake. I am trying to figure it out, inheriting from PrismApplication results in the WPF classes. CreateShell intends to return System.Windows.Window. How can add the base class to return Microsoft.UI.Xaml.Window?

**Reply by brianlagunas** [reply]
As mentioned in the beginning of this thread. Prism does not support WinUI. Only Prism.Uno can provide WinUI support via an Uno application.

**Reply by edoust** [reply]
This is not working for me, I needed to create a custom build of Prism that does not reference Uno.UI, but rather only WinUI directly and Uno.Core - otherwise it does not work for Uno + WinUI in a .NET 5 packaged desktop app.

Is this the right repo for that kind of issues? Is support for .NET 5 desktop packaged Uno + WinUI apps planned? The code already exists, to me it just seems to be a package referencing issue...

Do you have this on your to-do list, or would it be fine if I did a PR?

**Reply by sigmarsson** [reply]
Encountered exactly the same :

[Win UI + Prism](https://stackoverflow.com/questions/66572508/how-to-incorporate-win-ui-with-prism-across-the-uno-platform?noredirect=1#comment117685298_66572508)

**Reply by edoust** [reply]
> Encountered exactly the same :
> 
> [Win UI + Prism](https://stackoverflow.com/questions/66572508/how-to-incorporate-win-ui-with-prism-across-the-uno-platform?noredirect=1#comment117685298_66572508)

How did you solve it? Also, I encountered this error: https://github.com/unoplatform/uno/issues/5486

**Comment by tomasfabian** [comment]
Hi,
I have just installed Project reunion 0.5 into my Visual Studio and created a Blank app, Packaged (WinUI3 in Desktop). PrismApplication is not available in ```<PackageReference Include="Prism.Uno" Version="8.0.0.1909" />```.  

My MainWindow's type is Microsoft.UI.Xaml.Window. CreateShell expects Windows.UI.Xaml.UIElement:

```C#
protected override UIElement CreateShell()
```

I'm not sure whether this thread is about WinUI3 or WinUI, but my [original issue](https://github.com/unoplatform/Uno.Prism/issues/15) was closed.

**Reply by brianlagunas** [reply]
As mentioned earlier in this thread, WinUI is not supported. Also, you do realize Project Reunion 0.5 was just released yesterday. Even if Prism did support WinUI it wouldn't be updated and be available today with support for 0.5.

**Reply by edoust** [reply]
@brianlagunas Are there plans to support WinUI as part of ProjectReunion in the future? The missing support is preventing the usage of Prism in a .NET 5 desktop packaged app with WinUI (ProjectReunion)

@tomasfabian I created a ported version of PRISM that is working together with Project Reunion, I can share that with you if you like. It is only the DryIOC variant though. Basically you need to exchange the reference to Uno.UI to point to ProjectReunion, namespaces Microsoft and Windows need to be exchanged, and several other smaller issues in code have to be fixed in order to get it working. For me it is working without errors now

**Reply by brianlagunas** [reply]
@edoust Not at this time. The Prism team currently has no need to support WinUI. Since we have no need for WinUI, there is no reason for us to spend time on adding it and supporting it. If this changes, then that is when we will spend the time to add support for WinUI.

**Reply by tomasfabian** [reply]
@brianlagunas Roger that. I don't know why I was thinking that Prism was done for the community, not for the Prism team. I apologize for my mistake. Thank you anyway.

@edoust wow Sure you should share it with the community. You are very quick. Thanks to you, too!

**Reply by dansiegel** [reply]
it's worth remembering here... Uno Platform does have a vested interest and support for these things... So the support will likely come through Uno... what that looks like whether it just means you'll use the Prism.Uno package or whether there is some Prism.WinUI package supported by the Uno team is really TBD. Perhaps @jeromelaban can provide some useful insights here 😄

**Reply by brianlagunas** [reply]
@tomasfabian The community gets what they contribute. Dan and I contribute where we need the work. If the community needs other features, then the community can submit their work via a PR. That's how OSS works.

**Reply by tomasfabian** [reply]
@brianlagunas sure. Feature requests are also part of how OSS works. And honestly the situation was not clear who wants to do what. It looks like that @edoust has a solution, can you create a Pull request @edoust or share your solutions somehow please? I'm thinking about our own solution too, in case that no one contributes it. I suppose that Uno applications will have different XAML dialect ( I'm not familiar with Uno ).

**Reply by edoust** [reply]
@tomasfabian I can't create a PR, but I put my own version in [this fork](https://github.com/edoust/Prism/tree/feature/project-reunion). I am hosting the resulting Nuget packages on my private feed, seems to be working fine but with a huge startup delay like 10 seconds, that I couldn't figure out the reason for yet, if you find something out in that direction please let me know

**Reply by jeromelaban** [reply]
We're looking at creating version of Prism for WinUI/Reunion, and we were waiting on a stabilization of the API set (it changed between 0.5 preview and 0.5) and a name stabilization. Unfortunately, the `Microsoft.ProjectReunion` name will change again, and will break the packages that will have been created using it.

Also know that creating such a PR needs some additional nuget packaging and UI Testing work, not just the namespaces renames.

**Reply by 3D-aero-sbahr** [reply]
We are just starting a new project with uno and prism.
@jeromelaban Thank you for bringing the support for WinUI! The version of Prism for WinUI/Reunion you are creating will it support uno as well? What do you think when the support will be given. Days, weeks, months?

**Reply by jeromelaban** [reply]
At Uno Platform, we're generally tying the work we're contributing to the requests of our paying customers. It may happen sooner than later as WinUI gets more popular, yet we have no defined ETA at this point.

**Comment by dansiegel** [comment]
See PR #2393

**Comment by sigmarsson** [comment]
Currently I forged a set of IoC bindings as NInject Modules. Do you think I will be able to inject that entire module in the Prism IoC container - like AddModule() - or it will fusion with DryIoC/Unity only ? If the later is true, can i introduce such modules in any of these vendors' IoC?
